devdoc/waiter.md

```diff
@@ -0,0 +1,71 @@
+# `waiter` requirements
+
+## Overview
+
+`waiter` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers call `waiter_register_notification` to register a `NOTIFICATION_CALLBACK` function to be called when data becomes available.
+
+Producers call `waiter_notify` to notify the waiter that data is available.
+
+`waiter_register_notification` and `waiter_notify` can be called from different threads in any order.
```

Comment:
>order [](http://example.com/codeflow?start=95&length=5)

Does this mean "no concurrent calls"? What's the threading model here actually?

Reply:
Concurrent calls are ok.

`waiter` is thread-safe. The callbacks are called on the thread which makes the second call to `waiter_register_notification` or `waiter_notify`.